api/app/utils/s3.py

```diff
@@ -47,3 +48,13 @@ async def object_exists_v2(target_path: str):
     """
     async with get_client() as (client, bucket):
         return await object_exists(client, bucket, target_path)
+
+async def read_into_memory(key: str):
+    async with get_client() as (client, bucket):
+        s3_source = await client.get_object(Bucket=bucket, Key=key)
+        mem_path = f'/vsimem/{key}'
+        s3_data = await s3_source['Body'].read()
+        gdal.FileFromMemBuffer(mem_path, s3_data)
+        data = gdal.Open(mem_path, gdal.GA_ReadOnly)
```

Comment:
I think this call to `gdal.Open` is going to return a 'Source` object. We'll need a couple more steps to get the data in the format of an array.

If you want, you can also create a source object from S3 directly using GDAL, though it's not really any less verbose given that you have to set gdal config. See `classify_hfi.py` for an example.

Reply:
I think the tests pass because your raster_reader does the steps of getting the band and reading the data as an array.